packages/cli/src/services/hooks.service.ts

```diff
@@ -0,0 +1,93 @@
+import { Service } from 'typedi';
+import { UserService } from '@/services/user.service';
+import type { AssignableRole, User } from '@/databases/entities/User';
+import { AuthService } from '@/auth/auth.service';
+import type { Response } from 'express';
+import { UserRepository } from '@/databases/repositories/user.repository';
+import { SettingsRepository } from '@/databases/repositories/settings.repository';
+import type { QueryDeepPartialEntity } from '@n8n/typeorm/query-builder/QueryPartialEntity';
+import { WorkflowRepository } from '@/databases/repositories/workflow.repository';
+import { CredentialsRepository } from '@/databases/repositories/credentials.repository';
+import type { FindManyOptions, FindOneOptions, FindOptionsWhere } from '@n8n/typeorm';
+import type { WorkflowEntity } from '@/databases/entities/WorkflowEntity';
+import type { CredentialsEntity } from '@/databases/entities/CredentialsEntity';
+import type { Settings } from '@/databases/entities/Settings';
+
+/**
+ * Exposes functionality to be used by the cloud BE hooks.
+ * DO NOT DELETE or RENAME any of the methods without making sure this is not used in cloud BE hooks.
+ */
+@Service()
+export class HooksService {
+	constructor(
+		private userService: UserService,
+		private authService: AuthService,
+		private userRepository: UserRepository,
+		private settingsRepository: SettingsRepository,
+		private workflowRepository: WorkflowRepository,
+		private credentialsRepository: CredentialsRepository,
+	) {}
```

Comment:
Btw should we move the [dbCollections](https://github.com/n8n-io/n8n/blob/master/packages/cli/src/ExternalHooks.ts#L19) to this service as well? At the moment they are available in `this.dbCollections` in the hooks

Reply:
any changes to the external hooks interface should be considered a breaking change though, since we do have self-hosting customers who use external hooks.
we can definitely add the repositories here, but we should leave dbCollections as they are for now.

Reply:
@tomi, the main idea is to force us to use only methods we have in the hooks service. If the repositories are also available here, we might have a case where if we change the model for that repo, we will break the hooks and not know. However, if we only use the methods we have in the hooks, we will notice here that we broke them and adjusted the method to use the new format/change. Thoughts?

Reply:
It's always going to be balance between flexibility and runtime safety. I think it makes sense that certain things (like operations that change things in the DB or touch external systems like send email) have their own method to do that. But we can't think of every single use case beforehand, so thinking that we would do everything only based on the HooksService seems unrealistic. IMHO a good balance _could_ be to have the methods here that we know we need but also expose certain repositories so we also have the flexibility. We can then put everything we know we need here, and rely on the repositories to implement the functionality for older versions that don't yet have the new methods in the HooksService. So I'd expose the repositories here as well. Then it's a mutual agreement that we don't use those unless absolutely necessary and rely on tests (in the hooks side) to make sure they work for different n8n versions

Reply:
as long as we don't remove `dbCollections` from `ExternalHooks`, I agree with @tomi.
maybe we can mark it as `@deprecated` and remove it in v2.0.

Reply:
> But we can't think of every single use case beforehand, so thinking that we would do everything only based on the HooksService seems unrealistic. IMHO, a good balance could be to have the methods here that we know we need but also expose certain repositories so we also have the flexibility

I agree that specifying everything to be in the hooks is unrealistic, if we needed something, we would add it to the hooks service in n8n, wait for it to be release it, and then use it on the BE hooks via the hooks. So, we add functionality as we discover we need it.

>  We can then put everything we know we need here and rely on the repositories to implement the functionality for older versions that don't yet have the new methods in the HooksService.

For this we still have the `dbCollections ` we inject when loading the hooks. When people upgrade to a n8n version which include the hooks service then we use that instead. Once at instances have the hooks service we can stop using the `dbCollections ` altogether and as I mentioned before start adding functionality to the hooks as we need then. 

Conclusion:

In any case, we need some way to keep supporting access to these repositories because we have some self-hosting users relaying on this. So will:

1. Add the `dbCollections` here.
2. Mark the `dbCollections` in we pass as parameter in the external hooks as @deprecated